src/sagemaker/multidatamodel.py

```diff
@@ -0,0 +1,211 @@
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Placeholder docstring"""
+from __future__ import absolute_import
+
+import os
+from six.moves.urllib import parse
+
+import sagemaker
+from sagemaker.model import Model
+
+MULTI_MODEL_CONTAINER_MODE = "MultiModel"
+
+
+class MultiDataModel(Model):
+    """SageMaker ``MultiDataModel`` used to deploy multiple models to the same ``Endpoint``.
+
+    This class defines methods to create a Model with Multi-Model container, deploy it to an
+    endpoint, add more models to make them available to a deployed endpoint, list available
+    models and run predictions on a Multi-Model endpoint.
+    """
+
+    def __init__(
+        self,
+        model_name,
+        model_data_prefix,
+        model=None,
+        image=None,
+        role=None,
+        predictor_cls=None,
+        **kwargs
+    ):
+        """Initialize a ``MultiDataModel``. In addition to these arguments, it supports all
+           arguments supported by ``Model`` constructor
+
+        Args:
+            model_name (str): The model name.
+            model_data_prefix (str): The S3 prefix where all the models artifacts (.tar.gz)
+                in a Multi-Model endpoint are located
+            model (sagemaker.Model): The Model object that would define the
+                SageMaker model attributes like vpc_config, predictors, etc.
+                If this is present, the attributes for MultiDataModel are copied from this model
+                but are replaced the local arguments if present
+            image (str): A Docker image URI. If not provided, `model.image` is used instead.
+                (default: None)
+            role (str): An AWS IAM role (either name or full ARN). The Amazon
+                SageMaker training jobs and APIs that create Amazon SageMaker
+                endpoints use this role to access training data and model
+                artifacts. After the endpoint is created, the inference code
+                might use the IAM role if it needs to access some AWS resources.
+                It can be null if this is being used to create a Model to pass
+                to a ``PipelineModel`` which has its own Role field. If not provided,
+                `model.image` is used instead. (default: None)
+            predictor_cls (callable[string, sagemaker.session.Session]): A
+                function to call to create a predictor (default: None). If not
+                None, ``deploy`` will return the result of invoking this
+                function on the created endpoint name.
+            **kwargs: Keyword arguments passed to the ``Model`` initializer.
+        """
+        self.model_name = model_name
+        self._model_data_prefix = model_data_prefix
+        self.model_data_prefix = model_data_prefix
```

Comment:
What is the difference between the private and public `model_data_prefix` here?

Reply:
We needed the private variable as a helper for the validation that happens in the model_data_prefix.setter and self.model_data_prefix to actually set the value (which also triggers the setter and in the process validates the input). But in the new commit, I rewrote the logic to directly do the validation in init to avoid confusion